tests/nodeos_startup_catchup.py

```diff
@@ -0,0 +1,165 @@
+dumpErrorDetails=args.dump_error_details
+keepLogs=args.keep_logs
+dontKill=args.leave_running
+prodCount=args.prod_count if args.prod_count > 1 else 2
+killAll=args.clean_run
+p2pPlugin=args.p2p_plugin
+walletPort=args.wallet_port
+catchupCount=args.catchup_count if args.catchup_count > 0 else 1
+totalNodes=startedNonProdNodes+pnodes+catchupCount
+
+walletMgr=WalletMgr(True, port=walletPort)
+testSuccessful=False
+killEosInstances=not dontKill
+killWallet=not dontKill
+
+WalletdName=Utils.EosWalletName
+ClientName="cleos"
+
+try:
+    TestHelper.printSystemInfo("BEGIN")
+    cluster.setWalletMgr(walletMgr)
+
+    cluster.killall(allInstances=killAll)
+    cluster.cleanup()
+    specificExtraNodeosArgs={}
+    txnGenNodeNum=pnodes  # next node after producer nodes
+    for nodeNum in range(txnGenNodeNum, txnGenNodeNum+startedNonProdNodes):
+        specificExtraNodeosArgs[nodeNum]="--plugin eosio::txn_test_gen_plugin --txn-test-gen-account-prefix txntestacct"
+    Print("Stand up cluster")
+    if cluster.launch(prodCount=prodCount, onlyBios=False, pnodes=pnodes, totalNodes=totalNodes, totalProducers=pnodes*prodCount, p2pPlugin=p2pPlugin,
+                      useBiosBootFile=False, specificExtraNodeosArgs=specificExtraNodeosArgs, unstartedNodes=catchupCount, loadSystemContract=False) is False:
+        Utils.errorExit("Failed to stand up eos cluster.")
+
+    Print("Validating system accounts after bootstrap")
+    cluster.validateAccounts(None)
+
+    txnGenNodes=[]
+    for nodeNum in range(txnGenNodeNum, txnGenNodeNum+startedNonProdNodes):
+        txnGenNodes.append(cluster.getNode(nodeNum))
+
+    txnGenNodes[0].txnGenCreateTestAccounts(cluster.eosioAccount.name, cluster.eosioAccount.activePrivateKey)
+    time.sleep(20)
+
+    for genNum in range(0, len(txnGenNodes)):
+        salt="%d" % genNum
+        txnGenNodes[genNum].txnGenStart(salt, 1500, 150)
+        time.sleep(1)
+
+    node0=cluster.getNode(0)
+
+    def lib(node):
+        return node.getBlockNum(BlockType.lib)
+
+    def head(node):
+        return node.getBlockNum(BlockType.head)
+
+    time.sleep(10)
+    retryCountMax=100
+    for catchup_num in range(0, catchupCount):
+        lastLibNum=lib(node0)
+        lastHeadNum=head(node0)
+        lastCatchupLibNum=None
+
+        cluster.launchUnstarted(cachePopen=True)
+        retryCount=0
+        # verify that production node is advancing (sanity check)
+        while lib(node0)<=lastLibNum:
+            time.sleep(4)
+            retryCount+=1
+            # give it some more time if the head is still moving forward
+            if retryCount>=20 or head(node0)<=lastHeadNum:
+                Utils.errorExit("Node 0 failing to advance lib.  Was %s, now %s." % (lastLibNum, lib(node0)))
+            if Utils.Debug: Utils.Print("Node 0 head was %s, now %s.  Waiting for lib to advance" % (lastLibNum, lib(node0)))
+            lastHeadNum=head(node0)
+
+        catchupNode=cluster.getNodes()[-1]
+        time.sleep(9)
```

Comment:
In general the fewer the sleeps the better. I want to avoid depending on sleep time to avoid this test failing on slower/faster machines. Also we are in the process of trying to speed up tests.